nameko/messaging.py

```diff
@@ -105,30 +101,36 @@ def spam(self, data):
         self.exchange = exchange
         self.queue = queue
 
-    # TODO: should be a module level function
-    def get_connection(self):
-        conn = Connection(self.amqp_uri)
-        return connections[conn].acquire(block=True)
-
-    # TODO: should be a module level function
-    def get_producer(self):
-        conn = Connection(self.amqp_uri)
-        return producers[conn].acquire(block=True)
-
-    def setup(self):
+    @property
+    def amqp_uri(self):
+        return self.container.config[AMQP_URI_CONFIG_KEY]
 
-        self.amqp_uri = self.container.config[AMQP_URI_CONFIG_KEY]
+    @property
```

Comment:
`@property` makes them read-only on the instance, so you can only override them with a subclass. I _think_ that's a helpful restriction, but I could be persuaded otherwise.